server/src/metrics/client_metrics.rs

```diff
@@ -59,29 +88,118 @@ pub struct MetricsCache {
     pub metrics: DashMap<MetricsKey, ClientMetricsEnv>,
 }
 
+pub fn size_of_batch(batch: &MetricsBatch) -> usize {
+    serde_json::to_string(batch)
+        .map(|s| s.as_bytes().len())
+        .unwrap_or(0)
+}
+
+pub fn sendable(batch: &MetricsBatch) -> bool {
+    size_of_batch(batch) < UPSTREAM_MAX_BODY_SIZE
+}
+
+#[instrument(skip(batch))]
+pub fn cut_into_sendable_batches(batch: MetricsBatch) -> Vec<MetricsBatch> {
+    let batch_count = (size_of_batch(&batch) / BATCH_BODY_SIZE) + 1;
+    let apps_count = batch.applications.len();
+    let apps_per_batch = apps_count / batch_count;
+
+    let metrics_count = batch.metrics.len();
+    let metrics_per_batch = metrics_count / batch_count;
+
+    debug!("Batch count: {batch_count}. Apps per batch: {apps_per_batch}, Metrics per batch: {metrics_per_batch}");
+    (0..=batch_count)
+        .map(|counter| {
+            let apps_iter = batch.applications.iter();
+            let metrics_iter = batch.metrics.iter();
+            let apps_take = if apps_per_batch == 0 && counter == 0 {
+                apps_count
+            } else {
+                apps_per_batch
+            };
+            let metrics_take = if metrics_per_batch == 0 && counter == 0 {
+                metrics_count
+            } else {
+                metrics_per_batch
+            };
+            MetricsBatch {
+                metrics: metrics_iter
+                    .skip(counter * metrics_per_batch)
+                    .take(metrics_take)
+                    .cloned()
+                    .collect(),
+                applications: apps_iter
+                    .skip(counter * apps_per_batch)
+                    .take(apps_take)
+                    .cloned()
+                    .collect(),
+            }
+        })
+        .filter(|b| !b.applications.is_empty() || !b.metrics.is_empty())
+        .collect::<Vec<MetricsBatch>>()
+}
+
 impl MetricsCache {
-    pub fn get_unsent_metrics(&self) -> MetricsBatch {
-        MetricsBatch {
+    /// This is a destructive call. We'll remove all metrics that is due for posting
+    /// Called from [crate::http::background_send_metrics::send_metrics_task] which will reinsert on 5xx server failures, but leave 413 and 400 failures on the floor
+    pub fn get_appropriately_sized_batches(&self) -> Vec<MetricsBatch> {
+        let batch = MetricsBatch {
             applications: self
                 .applications
                 .iter()
                 .map(|e| e.value().clone())
                 .collect(),
-            metrics: self.metrics.iter().map(|e| e.value().clone()).collect(),
+            metrics: self
+                .metrics
+                .iter()
+                .map(|e| e.value().clone())
+                .filter(|m| m.yes > 0 || m.no > 0) // Makes sure that we only return buckets that have values. We should have a test for this :P
```

Comment:
We have a test for it!

```suggestion
                .filter(|m| m.yes > 0 || m.no > 0) // Makes sure that we only return buckets that have values
```